tools/deploy_s3.py

```diff
@@ -0,0 +1,124 @@
+        if e.response["Error"]["Code"] == "404":
+            return False
+
+        raise
+
+
+def _rm_s3_prefix(bucket: str, prefix: str):
+    """Remove all objects under a given prefix"""
+    s3 = boto3.resource("s3")
+    bucket = s3.Bucket(bucket)
+    for obj in bucket.objects.filter(Prefix=prefix):
+        obj.delete()
+
+
+@app.command()
+def deploy_to_s3_main(
+    local_folder: Path = typer.Argument(..., help="Path to the local folder"),
+    remote_prefix: Path = typer.Argument(..., help="Remote prefix"),
+    bucket: str = typer.Option(..., help="bucket name"),
+    cache_control: str = typer.Option(
+        "max-age=30758400, immutable, public", help="Cache control header to set"
+    ),
+    pretend: bool = typer.Option(False, help="Don't actually upload anything"),
+    overwrite: bool = typer.Option(False, help="Overwrite existing files"),
+    rm_remote_prefix: bool = typer.Option(
+        False, help="Remove existing files under the remote prefix"
+    ),
+):
+    """Deploy a dist folder with Pyodide packages to AWS S3"""
+    s3_client = boto3.client("s3")
+
+    typer.echo(f"Deploying {local_folder} to s3://{bucket}/{remote_prefix}")
+    typer.echo("Options: ")
+    typer.echo(f" - {cache_control=}")
+    typer.echo(f" - {pretend=}")
+    typer.echo(" - content-encoding: gzip")
+
+    if rm_remote_prefix:
+        _rm_s3_prefix(bucket, str(remote_prefix).lstrip("/"))
+
+    for file_path in local_folder.glob("**/*"):
+        if not file_path.is_file():
+            continue
+        remote_path = remote_prefix / file_path.relative_to(local_folder)
+
+        if not overwrite and check_s3_object_exists(
+            s3_client, bucket, str(remote_path).lstrip("/")
+        ):
+            typer.echo(
+                f"Cannot upload {file_path} because it already exists on s3://{bucket}/{remote_path}"
+            )
+            typer.echo("Exiting (use --overwrite to overwrite existing files)")
+            raise typer.Exit()
+
+        with open(file_path, "rb") as fh_in:
+            # Use gzip compression for storage. This only impacts storage on
+            # AWS and transfer between S3 and the CDN. It has no impact on the
+            # compression received by the end user (since the CDN re-compresses
+            # files).
+            fh_compressed = io.BytesIO()
+            with gzip.GzipFile(fileobj=fh_compressed, mode="w"):
+                shutil.copyfileobj(fh_in, fh_compressed)
+
+            fh_compressed.seek(0)
+
+            content_type = None
+            if file_path.suffix in (".zip", ".whl", ".tar", ".a"):
+                content_type = "application/wasm"
+            elif file_path.suffix == ".ts":
+                # This will not be correctly detected by mimetypes.
+                # However, JsDelivr will currently not serve .ts file in the
+                # custom CDN configuration, so it does not really matter.
+                content_type = "text/x.typescript"
```

Comment:
`.d.ts` files?

Reply:
Well the MIME type for .ts and .d.ts files is [a bit unclear](https://stackoverflow.com/a/13214026/1791279) but it doesn't really matter, since both are currently not allowed to be served via JsDelivr

https://cdn.jsdelivr.net/pyodide/v0.22.1/full/pyodide.ts (returns 403 forbidden)

So people get it from npm anyway.